File: LeetCode/98/main.py
"""
DIFFICULTY : medium
TAGS : tree, dfs, binary tree, binary search tree
"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    # Create a helper function for recursion
    
    # Use 2 global variables low and high to keep track of
    # the minimum and maximum value upto that node.  
    
    # The low and high update is simple : 
    # FOR LEFT SUBTREE : low = low ( no change ), high = root.val
    # FOR RIGHT SUBTREE : low = root.val, high = high ( no change )
    
    
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def traverse(root, low, high) :
            if not root :
                return True
            elif not low < root.val < high :
                return False
            else :
                return traverse(root.left, low, root.val) and traverse(root.right, root.val, high)
        return traverse(root, low=float("-inf"), high=float("inf"))
            

File: LeetCode/98/main.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func check(node *TreeNode, ans *bool) {
	if node == nil || !(*ans) {
		return
	}

	if (node.Left != nil && node.Val <= node.Left.Val) ||
		(node.Right != nil && node.Val >= node.Right.Val) {
		*ans = false
		return
	}

	check(node.Left, ans)
	check(node.Right, ans)

}

func isValidBST(root *TreeNode) bool {
	ans := true
	check(root, &ans)
	return ans
}
